File: Data Structures & Algorithms/sliding-window-maximum/submission-5.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        
        int n = nums.length;
        int l = 0, r = 0;
        ArrayDeque<Integer> dq = new ArrayDeque<>();
        int[] ans = new int[n-k+1];

        while( r < n)
        {
            //if top of dq is less than curr then no point in 
            //keeping in dq as they will never become max
            while(!dq.isEmpty() && nums[dq.getLast()] < nums[r])
                dq.removeLast();
            
            dq.addLast(r);

            if( l > dq.getFirst())
                dq.removeFirst();
            
            if( ( r + 1) >= k)
            {
                ans[l] = nums[dq.getFirst()];
                l++;
            }
            r++;
        }

        return ans;
    }
}

File: Data Structures & Algorithms/max-consecutive-ones-ii/submission-5.py
class Solution:
    def findMaxConsecutiveOnes(self, nums: List[int]) -> int:
        n = len(nums)
        l,r = 0,0
        ans,zc = 0,0

        while r < n:
            if nums[r] == 0:
                zc += 1
            
            while l <= r and zc > 1:
                if nums[l] == 0:
                    zc -= 1
                l += 1
            
            ans = max(ans,r-l+1)
            r += 1
        
        return ans